sage: 8.2 -> 8.3

timokau opened the pull request:
###### Motivation for this change

The first release candidate for 8.3 is out. It will probably still be some time until the proper release. Its easier to keep track on upstream changes by maintaining an unstable sage version. Since working in the open is always better and maybe somebody cares or wants to test/use it, I'm opening this WIP PR for it. If somebody thinks this shouldn't be a PR, feel free to close it.

Notes:

- the openblas fix is cherry-picked from #43234
-  the ipython patch will hopefully be in the latest upstream release by the time sage 8.3 is released

Known issues:

- transient test failure in `hyperelliptic_finite_field.py` https://groups.google.com/d/msg/sage-release/zSCoOwWazww/CtQqMXprBgAJ

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

timokau committed:
sage: 8.2 -> 8.3

timokau committed:
python2Packages.ipython: apply upstream signature patch

7c6f434c commented:
I think given the amount of complexity in upstream sage packaging it is a nice idea to have such a PR in the open.

FRidh commented on pkgs/development/libraries/science/math/openblas/default.nix:
```diff
@@ -117,7 +118,12 @@ stdenv.mkDerivation rec {
     ++ mapAttrsToList (var: val: var + "=" + val) config;
 
-  patches = []; # TODO: Remove on next mass-rebuild
+    patches = [
```
Why? Please add a comment.

timokau replied:
Thats the openblas backport I mentioned in the PR description, #43234

FRidh commented on pkgs/development/python-modules/ipython/5.nix:
```diff
@@ -38,6 +39,14 @@ buildPythonPackage rec {
   '';
 
+  patches = [
+    (fetchpatch {
```
Why? Please add a comment.

timokau replied:
Because sage applied that patch and relies on it. As I said in the PR description, I'm guessing/hoping there will be a new ipython release before the final sage 8.3. So that patch will likely never make it into nixpkgs master.

timokau replied:
Also see https://groups.google.com/forum/#!topic/sage-packaging/rUYd5M5tPTM

timokau commented:
> I think given the amount of complexity in upstream sage packaging it is a nice idea to have such a PR in the open.

Yah that's the idea :)

> PS. added some labels, found out that GitHub obviously cannot do a merge of independent label subset operations.

What do you mean by label subset operations? Race conditions with ofborg adding labels? (Not all that related, but I'm really looking forward to https://github.com/NixOS/ofborg/pull/164.)

timokau commented:
In case anybody is testing this and the tests fail, thats because of a transient test failure and not caused by the nix package: https://groups.google.com/d/msg/sage-release/zSCoOwWazww/CtQqMXprBgAJ

timokau commented:
> >What do you mean by label subset operations? Race conditions with ofborg adding labels?
> 
> Race condition between two disjoint additions!

Well thats just stupid.

> >(Not all that related, but I'm really looking forward to https://github.com/NixOS/ofborg/pull/164 .)
> 
> I look forward to Microsoft running GitHub into ground…
> 
> I mean, if we need ofborg for everything, maybe Github is a part of the problem, not part of a solution?

I wouldn't be all that saddened. But at gitlab wouldn't solve this problem either.
And since it seems like the move is not going to happen before the frog is boiled, we might as well make ourselves comfortable.

I dislike a lot of things about the sage development model. But the one thing I miss in nixpkgs is the workflow where every ticket has a status that is one of (among others) "needs:work" and "needs:review".

7c6f434c commented:
Well, Gitlab is better because you can have a NixOS test for ofborg-Gitlab integration! Of course, having a bot generate and upload summaries out of mailing list would probably work better than Github if only because scraping a mailing list enforces a proper model of operation.

I think _most_ of PRs are need:review by default, the rest are either [WIP] or match some needs: tag. And we don't have a clear idea what to do with issues, that's true. And of course we cannot even try a lot of nice workflows because «oh and run the full testsuite» requires insane amount of buildpower…

timokau commented:
> Well, Gitlab is better because you can have a NixOS test for ofborg-Gitlab integration! Of course, having a bot generate and upload summaries out of mailing list would probably work better than Github if only because scraping a mailing list enforces a proper model of operation.

What do you mean by generating summaries out of mailinglists?

> I think _most_ of PRs are need:review by default, the rest are either [WIP] or match some needs: tag. And we don't have a clear idea what to do with issues, that's true.

Well they all *start* with needs:review, but I think many are in the "needs:work" phase of the review. If we had consistent tagging reviewers could just look at the needs:review list when they have time and we could properly judge how bad our PR-problem is.

> And of course we cannot even try a lot of nice workflows because «oh and run the full testsuite» requires insane amount of buildpower…

Not sure what you mean by that. The full ofBorg testsuite?

7c6f434c commented:
>What do you mean by generating summaries out of mailinglists?

Patches to mailing list sometimes get lost. A summary «we have this Git URL submitted as a pull request, these people approved, these are ofborg evaluation result» and a global overview of such summaries would probably become strictly more informative than Github after three months.

>Well they all *start* with needs:review, but I think many are in the "needs:work" phase of the review. If we had consistent tagging reviewers could just look at the needs:review list when they have time and we could properly judge how bad our PR-problem is.

Dismissing changes-requested reviews when no longer applicable could also be an option… And WIP labels are used from time to time. I would guess that the PR-problem is worse than you think, though. Then again, there is a nontrivial number of stuck PRs by project members (which are almost always a completely different kind of problem). Well, if we used an actual issue tracker we would have separate trackers for separate things, but oh well.

>Not sure what you mean by that. The full ofBorg testsuite?

No, checking that a minor package update doesn't break any reverse dependencies.

timokau commented:
On 18-07-11 14:05, Michael Raskin wrote:
> >What do you mean by generating summaries out of mailinglists?
> 
> Patches to mailing list sometimes get lost. A summary «we have this Git URL submitted as a pull request, these people approved, these are ofborg evaluation result» and a global overview of such summaries would probably become strictly more informative than Github after three months.

Ah. Not sure what I think of the mailing list approach. It does have its merits for sure.

> >Well they all *start* with needs:review, but I think many are in the "needs:work" phase of the review. If we had consistent tagging reviewers could just look at the needs:review list when they have time and we could properly judge how bad our PR-problem is.
> 
> Dismissing changes-requested reviews when no longer applicable could also be an option…

Yeah that has always bothered me.

> And WIP labels are used from time to time.

But usually not in the normal back-and-forth between reviewer and contributor.

> I would guess that the PR-problem is worse than you think, though. Then again, there is a nontrivial number of stuck PRs by project members (which are almost always a completely different kind of problem).

Yeah the non-trivial, infrastructure changing PRs are something different entirely.

> Well, if we used an actual issue tracker we would have separate trackers for separate things, but oh well.

I prefer tagging over separate trackers, but thats basically the same thing.


By the way could someone merge #43234? This is causing failures and considering that it'll have to go through staging I would like to merge that as quickly as possible. The patch is already upstream and it is a bug fix so that should be pretty uncontroversial.

timokau commented:
> No, checking that a minor package update doesn't break any reverse dependencies.

Ah right. But that shouldn't stop us from improving our current workflow.

By the way in relevant sage related news: There is [work upstream](https://trac.sagemath.org/ticket/25270) to make it possible to run tests in limited time. That would make it possible to test most of sage in a fraction of the time. I was thinking about completely splitting the tests from the main package anyways (while still making them a dependency). So then there could be `sage.tests` (which would be a dependency of the sage package) and `sage.quicktests` which would run maybe 10 minutes and could be used for reverse-dependency checks.

7c6f434c commented:
Good news for Sage tests, but Sage is still a non-negligible rebuild per se, and then it is not the only one.

As for workflow — some structural differences mean that many tradeoffs between using tooling and spending human attention become unaffodable to us, and handling Nixpkgs PRs en masse is a burn out risk when spending too much of  human attention.

But of course, there are a lot of good ideas that do not require making extra decisions or ensuring that nobody forgets about a PR.

timokau committed:
singular: backport ntl error handling patch

timokau committed:
sage: don't pin singular

timokau committed:
sage: don't pin pari

timokau committed:
lcalc: apply c++11 patch

timokau committed:
sage: don't pin ntl

Now possible because of the lcalc c++11 patch.

timokau commented:
8.3 was released :tada: I've rebased and squashed the intermediate rc updates. The final build is still running but in the meanwhile:

@bjornfor @jgeerds @FRidh: Unfortunately there was no new ipython release. Is it okay to backport that patch in https://github.com/NixOS/nixpkgs/pull/43281/commits/86f058c8427da6a563b799aecf6c4cccdd395278? It has been accepted upstream.

FRidh commented:
@timokau I think so. Note I don't maintain the Python 2 version of IPython.

timokau commented:
I've added a patch to mark a known "random" failure (always happens on my laptop, never on my desktop). This is ready for merge.
